### PR TITLE
More symbols to singleton

### PR DESCRIPTION
**Pull request by mmatera**

### CODE REVIEW

File: mathics/builtin/numbers/calculus.py [hunk 1]
```diff
@@ -1449,9 +1451,7 @@ def apply(self, f, x, x0, evaluation, options):
             method = method.value
 
         # Determine the "jacobian"
-        if method in ("Newton",) and options["System`Jacobian"].sameQ(
-            Symbol("Automatic")
-        ):
+        if method in ("Newton",) and options["System`Jacobian"].sameQ(SymbolAutomatic):
```
Comment (TiagoCavalcante) [inline mathics/builtin/numbers/calculus.py:1454]: `in ("Newton",)` -> `== "Newton"`

File: mathics/core/definitions.py [hunk 2]
```diff
@@ -494,7 +498,13 @@ def get_definition(self, name, only_if_exists=False) -> "Definition":
         return definition
 
     def get_attributes(self, name):
-        return self.get_definition(name).attributes
+        definition = self.definitions_cache.get(name, None)
```
Comment (TiagoCavalcante) [inline mathics/core/definitions.py:501]: @mmatera I think that caching is not necessary here, it is just acessing a class variable.

File: mathics/core/definitions.py [hunk 3]
```diff
@@ -578,6 +588,8 @@ def set_attributes(self, name, attributes) -> None:
         self.clear_definitions_cache(name)
 
     def clear_attribute(self, name, attribute) -> None:
+        if isinstance(attribute, str):
```
Comment (TiagoCavalcante) [inline mathics/core/definitions.py:591]: I also think this is not necessary.

File: mathics/core/expression.py [hunk 4]
```diff
@@ -1297,6 +1323,7 @@ def thread(self, evaluation, head=None) -> typing.Tuple[bool, "Expression"]:
             return True, Expression(head, *leaves)
 
     def is_numeric(self, evaluation=None) -> bool:
+        name = self._head.get_name()
```
Comment (TiagoCavalcante) [inline mathics/core/expression.py:1326]: I'm not seeing where this variable is used.

File: mathics/core/pattern.py [hunk 5]
```diff
@@ -2,11 +2,12 @@
 # cython: profile=False
 # -*- coding: utf-8 -*-
 
-
+from mathics.core.symbols import Symbol
```
Comment (TiagoCavalcante) [inline mathics/core/pattern.py:5]: Also not seeing where this is used.

File: mathics/core/pattern.py [hunk 6]
```diff
@@ -2,11 +2,12 @@
...
 from mathics.core.expression import Expression
 from mathics.core.symbols import system_symbols, ensure_context, Atom, Symbol
 from mathics.core.util import subsets, subranges, permutations
 from itertools import chain
+from mathics.core.systemsymbols import SymbolFlat, SymbolOrderless, SymbolOneIdentity
```
Comment (TiagoCavalcante) [inline mathics/core/pattern.py:10]: Neither this.

File: mathics/core/symbols.py [hunk 7]
```diff
@@ -707,7 +712,12 @@ def default_format(self, evaluation, form) -> str:
         return self.name
 
     def get_attributes(self, definitions):
-        return definitions.get_attributes(self.name)
+        name = self.name
```
Comment (TiagoCavalcante) [inline mathics/core/symbols.py:715]: I also think caching is not necessary here.

File: mathics/builtin/assignments/internals.py [hunk 8]
```diff
@@ -587,44 +597,44 @@ def process_tags_and_upset_allow_custom(tags, upset, self, lhs, evaluation):
...
-        if not name:
+        head = focus.get_lookup_symbol()
+        if not head:
             evaluation.message(self.get_name(), "setraw", focus)
             raise AssignmentException(lhs, None)
-        tags = [name]
+        tags = [head]
     elif upset:
         tags = []
         if focus.is_atom():
```
Comment (TiagoCavalcante) [inline mathics/builtin/assignments/internals.py:613]: I think `is_symbol` and `is_atom` should be class variables.

**Review by TiagoCavalcante — COMMENTED** [review]
Overall looks great, I'll check if there are more `Expression("Head", ...)` to convert to `Expression(SymbolHead, ...)` tomorrow.

File: mathics/builtin/numbers/calculus.py (hunk 1 above)
Comment (mmatera) [reply mathics/builtin/numbers/calculus.py:1454]: Yes, all this needs more work. Y just tried to have a working branch to see the differences and try to improve it.

File: mathics/core/definitions.py (hunk 2 above)
Comment (mmatera) [reply mathics/core/definitions.py:501]: @TiagoCavalcante, this is an experiment: function calls in Python take 2 to 3 times the time required to access a dictionary by its key. Also, most of the times, (really, 20000 to 1 during doctest) the definition is in the cache, so the call is not necessary.   My hypothesis then is that avoiding calls to get_definition()  when they are not needed could speed up in a constant factor the evaluation. But again, this is just an experiment...

Comment (TiagoCavalcante) [reply mathics/core/definitions.py:501]: @mmatera I'll make some changes tomorrow and put it in another PR. I think this worth dedicated benchmarks.

**Comment by TiagoCavalcante** [comment]
@mmatera thinking well, all of this should be targetting master, not faster-attributes-bitset, this is not related to attributes-bitset, what do you think?

**Comment by TiagoCavalcante** [comment]
@mmatera @rocky I'm going to rebase this and do the changes I mentioned after attributes-bitset be merged.

**Comment by rocky** [comment]
rebasing and squashing commits is always appreciated.

File: mathics/core/pattern.py (hunk 6 above)
Comment (mmatera) [reply mathics/core/pattern.py:10]: Now used anymore because now we use the bitmask

**Comment by mmatera** [comment]
deprecated

**Pull request closed without merging**